drivers/rtc/rtc_stm32.c

```diff
@@ -0,0 +1,322 @@
+	RTC_AlarmStructure.AlarmMask
+		= RTC_ALARMMASK_DATEWEEKDAY|RTC_ALARMMASK_HOURS;
+	RTC_AlarmStructure.AlarmSubSecondMask   = RTC_ALARMSUBSECONDMASK_NONE;
+	result = HAL_RTC_SetAlarm_IT(&context->hrtc,
+				     &RTC_AlarmStructure,
+				     RTC_FORMAT_BIN);
+	if (result) {
+		result = -EIO;
+	}
+	return result;
+}
+
+static u32_t rtc_stm32_read(struct device *dev)
+{
+	struct rtc_stm32_data *context = dev->driver_data;
+	RTC_TimeTypeDef sTime;
+	RTC_DateTypeDef sDate;
+	u32_t TimeMs;
+	int result;
+
+	result = HAL_RTC_GetDate(&context->hrtc, &sDate, RTC_FORMAT_BIN);
+	if (result) {
+		return -EIO;
+	}
+	result = HAL_RTC_GetTime(&context->hrtc, &sTime, RTC_FORMAT_BIN);
+	if (result) {
+		return -EIO;
+	}
+
+	TimeMs = (60*1000*sTime.Minutes)+(1000*sTime.Seconds) +
+		(1000*sTime.SubSeconds/(sTime.SecondFraction+1));
+
+	return TimeMs;
+}
+
+static u32_t rtc_stm32_get_pending_int(struct device *dev)
+{
+	/* Not available for STM32 RTC */
+	return 0;
+}
+#ifdef CONFIG_SOC_SERIES_STM32F4X
+void rtc_stm32_isr(int line, void *userdata)
+{
+	struct device * const dev = (struct device *)userdata;
+	struct rtc_stm32_data *context = dev->driver_data;
+
+	HAL_RTC_AlarmIRQHandler(&context->hrtc);
+	rtc_data.cb_fn(dev);
+}
+#endif
+#ifdef CONFIG_SOC_SERIES_STM32L4X
+void rtc_stm32_isr(void *userdata)
+{
+	struct device * const dev = (struct device *)userdata;
+	struct rtc_stm32_data *context = dev->driver_data;
+
+	HAL_RTC_AlarmIRQHandler(&context->hrtc);
+	rtc_data.cb_fn(dev);
+}
+#endif
+static const struct rtc_driver_api rtc_api = {
+		.enable = rtc_stm32_enable,
+		.disable = rtc_stm32_disable,
+		.read = rtc_stm32_read,
+		.set_config = rtc_stm32_set_config,
+		.set_alarm = rtc_stm32_set_alarm,
+		.get_pending_int = rtc_stm32_get_pending_int,
+};
+
+#ifdef CONFIG_DEVICE_POWER_MANAGEMENT
+
+#ifdef CONFIG_SYS_POWER_DEEP_SLEEP
+static int rtc_suspend_device(struct device *dev)
+{
+	__HAL_RCC_RTC_DISABLE();
+	rtc_stm32_set_power_state(dev, DEVICE_PM_SUSPEND_STATE);
+	return 0;
+}
+
+static int rtc_resume_device(struct device *dev)
+{
+	__HAL_RCC_RTC_ENABLE();
+	rtc_stm32_set_power_state(dev, DEVICE_PM_ACTIVE_STATE);
+
+	return 0;
+}
+#endif
+
+/*
+ * Implements the driver control management functionality
+ * the *context may include IN data or/and OUT data
+ */
+static int rtc_stm32_device_ctrl(struct device *dev, u32_t ctrl_command,
+				 void *context)
+{
+	if (ctrl_command == DEVICE_PM_SET_POWER_STATE) {
+#ifdef CONFIG_SYS_POWER_DEEP_SLEEP
+		if (*((u32_t *)context) == DEVICE_PM_SUSPEND_STATE) {
+			return rtc_suspend_device(dev);
+		} else if (*((u32_t *)context) == DEVICE_PM_ACTIVE_STATE) {
+			return rtc_resume_device(dev);
+		}
+#endif
+	} else if (ctrl_command == DEVICE_PM_GET_POWER_STATE) {
+		*((u32_t *)context) = rtc_stm32_get_power_state(dev);
+		return 0;
+	}
+
+	return 0;
+}
+#endif
+static int rtc_stm32_init(struct device *dev);
+DEVICE_DEFINE(rtc_stm32, CONFIG_RTC_0_NAME,
```

Comment:
move that macro call at the end of the file (and you won't need to declare rtc_stm32_init prematurely as well)

Reply:
Please do the same for _data and _config structures

Reply:
Done